From e23e8671ab4e7af720bd82c13c31f1c4182328d3 Mon Sep 17 00:00:00 2001
From: Leon Leijssen <leon.leijssen@topic.nl>
Date: Tue, 25 Feb 2020 17:25:24 +0100
Subject: [PATCH 15/34] topic-miami: support new filesystem structure

Add MTD and UBI support and boot from QSPI using UBI filesystem
---
 arch/arm/dts/zynq-topic-miami.dts | 16 +++------
 configs/topic_miami_defconfig     | 17 ++++++++++
 configs/topic_miamilite_defconfig | 18 ++++++++++
 configs/topic_miamiplus_defconfig | 18 ++++++++++
 include/configs/topic_miami.h     | 56 ++++++++++++++++++++++---------
 5 files changed, 98 insertions(+), 27 deletions(-)

diff --git a/arch/arm/dts/zynq-topic-miami.dts b/arch/arm/dts/zynq-topic-miami.dts
index e50ae58a23..463eab21f0 100644
--- a/arch/arm/dts/zynq-topic-miami.dts
+++ b/arch/arm/dts/zynq-topic-miami.dts
@@ -50,23 +50,15 @@
 		};
 		partition@10000 {
 			label = "qspi-u-boot-img";
-			reg = <0x20000 0x80000>;
+			reg = <0x20000 0xC0000>;
 		};
-		partition@qspi-u-boot-env {
+		partition@A0000 {
 			label = "qspi-u-boot-env";
 			reg = <0xA0000 0x20000>;
 		};
-		partition@70000 {
-			label = "qspi-device-tree";
-			reg = <0xC0000 0x20000>;
-		};
-		partition@80000 {
-			label = "qspi-linux";
-			reg = <0x100000 0x480000>;
-		};
-		partition@480000 {
+		partition@100000 {
 			label = "qspi-rootfs";
-			reg = <0x580000 0x1A80000>;
+			reg = <0x100000 0x1F00000>;
 		};
 	};
 };
diff --git a/configs/topic_miami_defconfig b/configs/topic_miami_defconfig
index 708a05f74f..350b9ac74e 100644
--- a/configs/topic_miami_defconfig
+++ b/configs/topic_miami_defconfig
@@ -28,6 +28,12 @@ CONFIG_CMD_DFU=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
+CONFIG_CMD_MTD=y
+CONFIG_CMD_MTDPARTS=y
+# CONFIG_CMD_MTDPARTS_SPREAD is not set
+CONFIG_CMD_SF=y
+CONFIG_CMD_UBI=y
+CONFIG_CMD_UBIFS=y
 CONFIG_CMD_USB=y
 # CONFIG_CMD_SETEXPR is not set
 # CONFIG_CMD_NET is not set
@@ -47,6 +53,8 @@ CONFIG_DM_I2C=y
 CONFIG_SYS_I2C_CADENCE=y
 CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ZYNQ=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_BAR=y
 CONFIG_SF_DEFAULT_MODE=0
 CONFIG_SF_DEFAULT_SPEED=108000000
 CONFIG_SPI_FLASH_STMICRO=y
@@ -66,3 +74,12 @@ CONFIG_USB_GADGET_PRODUCT_NUM=0x0300
 CONFIG_CI_UDC=y
 CONFIG_USB_GADGET_DOWNLOAD=y
 CONFIG_USB_FUNCTION_THOR=y
+CONFIG_MTDIDS_DEFAULT="nor0=spi0.0"
+CONFIG_MTDPARTS_DEFAULT="mtdparts=spi0.0:0x20000(qspi-u-boot-spl),0xC0000(qspi-u-boot-img),0x20000(qspi-u-boot-env),-(qspi-rootfs)"
+# CONFIG_DFU_NAND is not set
+# CONFIG_FDT_FIXUP_PARTITIONS is not set
+CONFIG_LZO=y
+CONFIG_MTD=y
+# CONFIG_CFI_FLASH is not set
+# CONFIG_ALTERA_QSPI is not set
+# CONFIG_MTD_SPI_NAND is not set
diff --git a/configs/topic_miamilite_defconfig b/configs/topic_miamilite_defconfig
index 45c2f93ae9..e2a977226f 100644
--- a/configs/topic_miamilite_defconfig
+++ b/configs/topic_miamilite_defconfig
@@ -26,6 +26,12 @@ CONFIG_CMD_DFU=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
+CONFIG_CMD_MTD=y
+CONFIG_CMD_MTDPARTS=y
+# CONFIG_CMD_MTDPARTS_SPREAD is not set
+CONFIG_CMD_SF=y
+CONFIG_CMD_UBI=y
+CONFIG_CMD_UBIFS=y
 CONFIG_CMD_USB=y
 # CONFIG_CMD_SETEXPR is not set
 # CONFIG_CMD_NET is not set
@@ -42,6 +48,9 @@ CONFIG_DM_I2C=y
 CONFIG_SYS_I2C_CADENCE=y
 CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ZYNQ=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_BAR=y
+CONFIG_SF_DUAL_FLASH=y
 CONFIG_SF_DEFAULT_MODE=0
 CONFIG_SF_DEFAULT_SPEED=108000000
 CONFIG_SPI_FLASH_STMICRO=y
@@ -61,3 +70,12 @@ CONFIG_USB_GADGET_PRODUCT_NUM=0x0300
 CONFIG_CI_UDC=y
 CONFIG_USB_GADGET_DOWNLOAD=y
 CONFIG_USB_FUNCTION_THOR=y
+CONFIG_MTDIDS_DEFAULT="nor0=spi0.0"
+CONFIG_MTDPARTS_DEFAULT="mtdparts=spi0.0:0x20000(qspi-u-boot-spl),0xC0000(qspi-u-boot-img),0x20000(qspi-u-boot-env),-(qspi-rootfs)"
+# CONFIG_DFU_NAND is not set
+# CONFIG_FDT_FIXUP_PARTITIONS is not set
+CONFIG_LZO=y
+CONFIG_MTD=y
+# CONFIG_CFI_FLASH is not set
+# CONFIG_ALTERA_QSPI is not set
+# CONFIG_MTD_SPI_NAND is not set
diff --git a/configs/topic_miamiplus_defconfig b/configs/topic_miamiplus_defconfig
index 44481cf1cf..96de2eb501 100644
--- a/configs/topic_miamiplus_defconfig
+++ b/configs/topic_miamiplus_defconfig
@@ -26,6 +26,12 @@ CONFIG_CMD_DFU=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
+CONFIG_CMD_MTD=y
+CONFIG_CMD_MTDPARTS=y
+# CONFIG_CMD_MTDPARTS_SPREAD is not set
+CONFIG_CMD_SF=y
+CONFIG_CMD_UBI=y
+CONFIG_CMD_UBIFS=y
 CONFIG_CMD_USB=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_CMD_CACHE=y
@@ -41,6 +47,9 @@ CONFIG_DM_I2C=y
 CONFIG_SYS_I2C_CADENCE=y
 CONFIG_MMC_SDHCI=y
 CONFIG_MMC_SDHCI_ZYNQ=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_BAR=y
+CONFIG_SF_DUAL_FLASH=y
 CONFIG_SF_DEFAULT_MODE=0
 CONFIG_SF_DEFAULT_SPEED=108000000
 CONFIG_SPI_FLASH_STMICRO=y
@@ -61,3 +70,12 @@ CONFIG_USB_GADGET_PRODUCT_NUM=0x0300
 CONFIG_CI_UDC=y
 CONFIG_USB_GADGET_DOWNLOAD=y
 CONFIG_USB_FUNCTION_THOR=y
+CONFIG_MTDIDS_DEFAULT="nor0=spi0.0"
+CONFIG_MTDPARTS_DEFAULT="mtdparts=spi0.0:0x20000(qspi-u-boot-spl),0xC0000(qspi-u-boot-img),0x20000(qspi-u-boot-env),-(qspi-rootfs)"
+# CONFIG_DFU_NAND is not set
+# CONFIG_FDT_FIXUP_PARTITIONS is not set
+CONFIG_LZO=y
+CONFIG_MTD=y
+# CONFIG_CFI_FLASH is not set
+# CONFIG_ALTERA_QSPI is not set
+# CONFIG_MTD_SPI_NAND is not set
diff --git a/include/configs/topic_miami.h b/include/configs/topic_miami.h
index bd59532bba..669ecab26c 100644
--- a/include/configs/topic_miami.h
+++ b/include/configs/topic_miami.h
@@ -44,6 +44,14 @@
 /* Faster flash, ours may run at 108 MHz */
 #undef CONFIG_SPI_FLASH_WINBOND
 
+/* CONFIG_SPI_FLASH_MTD only works outside SPL */
+#if !defined(CONFIG_SPL_BUILD)
+# if defined(CONFIG_ZYNQ_QSPI)
+#  define CONFIG_SPI_FLASH_MTD
+# endif
+#endif
+
+
 /* Setup proper boot sequences for Miami boards */
 
 #if defined(CONFIG_USB)
@@ -68,7 +76,7 @@
 # define DFU_ALT_INFO \
 	"dfu_alt_info=" \
 	"uImage ram 0x2080000 0x500000;" \
-	"devicetree.dtb ram 0x2000000 0x20000;" \
+	"system.dtb ram 0x2000000 0x20000;" \
 	"uramdisk.image.gz ram 0x4000000 0x10000000\0" \
 	"dfu_ram=run usbreset && dfu 0 ram 0\0" \
 	"thor_ram=run usbreset && thordown 0 ram 0\0"
@@ -78,19 +86,40 @@
 
 #undef CONFIG_EXTRA_ENV_SETTINGS
 #define CONFIG_EXTRA_ENV_SETTINGS	\
+	"bootargs=quiet\0" \
+	"boot_a_script=load ${devtype} ${devnum}:${distro_bootpart} ${scriptaddr} ${prefix}${script}; source ${scriptaddr}\0" \
+	"boot_prefixes=/ /boot/\0" \
+	"boot_scripts=boot.scr\0" \
+	"boot_targets=jtag mmc0 qspi0 usb0 usb1\0" \
+	"bootcmd_jtag=source $scriptaddr; echo SCRIPT FAILED: continuing...;\0" \
+	"bootcmd_mmc0=setenv devnum 0; run mmc_boot\0" \
+	"bootcmd_mmc1=setenv devnum 1; run mmc_boot\0" \
+	"bootcmd_qspi=sf probe 0 && run bootcmd_ubi\0" \
+	"bootcmd_ubi=run ubi_mount && ubifsload ${scriptaddr} /boot/boot.scr && source ${scriptaddr}\0" \
+	"bootcmd_usb0=setenv devnum 0; run usb_boot\0" \
+	"bootcmd_usb1=setenv devnum 1; run usb_boot\0" \
+	"distro_bootcmd=for target in ${boot_targets}; do run bootcmd_${target}; done\0" \
 	"kernel_image=uImage\0"	\
 	"kernel_addr=0x2080000\0" \
+	"mmc_boot=if mmc dev ${devnum}; then setenv devtype mmc; run scan_dev_for_boot_part; fi\0" \
 	"ramdisk_image=uramdisk.image.gz\0"	\
 	"ramdisk_load_address=0x4000000\0"	\
-	"devicetree_image=devicetree.dtb\0"	\
+	"devicetree_image=system.dtb\0"	\
 	"devicetree_addr=0x2000000\0"	\
 	"bitstream_image=fpga.bin\0"	\
-	"bootscript=autorun.scr\0" \
-	"loadbit_addr=0x100000\0"	\
+	"bootscript=boot.scr\0" \
+	"scriptaddr=0x1900000\0" \
+	"loadbit_addr=0x100000\0" \
+	"scan_dev_for_boot=echo Scanning ${devtype} ${devnum}:${distro_bootpart}...; for prefix in ${boot_prefixes}; do run scan_dev_for_scripts; done\0" \
+	"scan_dev_for_boot_part=part list ${devtype} ${devnum} -bootable devplist; env exists devplist || setenv devplist 1; for distro_bootpart in ${devplist}; do if fstype ${devtype} ${devnum}:${distro_bootpart} bootfstype; then run scan_dev_for_boot; fi; done\0" \
+	"scan_dev_for_scripts=for script in ${boot_scripts}; do if test -e ${devtype} ${devnum}:${distro_bootpart} ${prefix}${script}; then echo Found U-Boot script ${prefix}${script}; run boot_a_script; echo SCRIPT FAILED: continuing...; fi; done\0" \
 	"loadbootenv_addr=0x2000000\0" \
+	"mtdids=" CONFIG_MTDIDS_DEFAULT "\0" \
+	"mtdparts=" CONFIG_MTDPARTS_DEFAULT "\0" \
 	"kernel_size=0x480000\0"	\
 	"devicetree_size=0x10000\0"	\
 	"boot_size=0xF00000\0"	\
+	"fdt_addr=4000000 \0" \
 	"fdt_high=0x20000000\0"	\
 	"initrd_high=0x20000000\0"	\
 	"mmc_loadbit=echo Loading bitstream from SD/MMC/eMMC to RAM.. && " \
@@ -102,21 +131,18 @@
 		"sf read ${devicetree_addr} 0xC0000 ${devicetree_size} && " \
 		"sf read ${kernel_addr} 0x100000 ${kernel_size} && " \
 		"bootm ${kernel_addr} - ${devicetree_addr}\0" \
-	"sdboot=if mmcinfo; then " \
-			"setenv bootargs console=ttyPS0,115200 " \
-				"root=/dev/mmcblk0p2 rw rootfstype=ext4 " \
-				"rootwait quiet ; " \
-			"load mmc 0 ${kernel_addr} ${kernel_image}&& " \
-			"load mmc 0 ${devicetree_addr} ${devicetree_image}&& " \
-			"bootm ${kernel_addr} - ${devicetree_addr}; " \
-		"fi\0" \
+	"sdboot=mmc dev $sdbootdev && mmcinfo && if load mmc $sdbootdev:$partid ${scriptaddr} boot.scr; then source ${scriptaddr}; fi; run sdroot$sdbootdev; load mmc $sdbootdev:$partid $fdt_addr system.dtb && load mmc $sdbootdev:$partid $kernel_addr Image && booti $kernel_addr - $fdt_addr\0" \
+	"sdbootdev=0\0" \
+	"sdroot0=setenv bootargs $bootargs root=/dev/mmcblk0p2 rw rootwait\0" \
+	"sdroot1=setenv bootargs $bootargs root=/dev/mmcblk1p2 rw rootwait\0" \
+	"ubi_mount=sf probe && ubi part qspi-rootfs && ubifsmount ubi0:qspi-rootfs && setenv devtype ubi\0" \
+	"bootubipart=qspi-rootfs\0" \
+	"bootubivol=qspi-rootfs\0" \
 	EXTRA_ENV_USB \
 	DFU_ALT_INFO
 
 #undef CONFIG_BOOTCOMMAND
-#define CONFIG_BOOTCOMMAND	"if mmcinfo; then " \
-	"if fatload mmc 0 0x1900000 ${bootscript}; then source 0x1900000; " \
-	"fi; fi; run $modeboot"
+#define CONFIG_BOOTCOMMAND	"run distro_bootcmd"
 #undef CONFIG_DISPLAY_BOARDINFO
 
 #endif /* __CONFIG_TOPIC_MIAMI_H */
-- 
2.17.1

